### ISSUE TITLE
input should not require number generated always as identity

### ISSUE THREAD

**Issue by wenzowski** [graphile/postgraphile#631]
I'm submitting a ...
- [x] bug report
- [ ] feature request
- [ ] question

PostGraphQL version:
`4.0.0-alpha2.22`

PostgreSQL version:
`psql (PostgreSQL) 10.0` for [identity columns](https://blog.2ndquadrant.com/postgresql-10-identity-columns/)

Minimal SQL file that can be loaded into a clean database:
``` sql
CREATE TABLE test_new (
    id bigint PRIMARY KEY,
    number bigint GENERATED ALWAYS AS IDENTITY,
);
```

### Current behavior:
```graphql
type TestNewInput implements Node {
  id: UUID
  number: BigInt!
}
```
This is surprising because [postgres docs say](https://www.postgresql.org/docs/10/static/sql-createtable.html)
> If ALWAYS is specified, a user-specified value is only accepted if the INSERT statement specifies OVERRIDING SYSTEM VALUE. If BY DEFAULT is specified, then the user-specified value takes precedence.
Identity columns must be `NOT NULL`, so this is the expected behaviour for `BY DEFAULT` but not for `ALWAYS`.

### Expected behavior:
```graphql
type TestNewInput implements Node {
  id: UUID
}
```
This is desirable because any user-specified value for the `number` column will be dropped by default.

**Comment by benjie** [comment]
This... This is a pain. You're 100% right; but we haven't added any specific PG10 support to PostGraphile yet. The identity stuff adds the `attidentity` column to [pg_attribute](https://www.postgresql.org/docs/10/static/catalog-pg-attribute.html) which we could totally factor into our decision making; however we can't simply add that to the introspection query because it will not run on PG<10.

So here's the options I see at first blush:

1. have two introspection queries, one for PG10 and the other for PG9 and run the relevant one. Ugh; maintenance overhead. Not keen on this.
2. have two introspection queries that are *generated* from one file, so you only maintain one file and the PG10 stuff is wrapped in something that'll result in it being removed from the PG9 query.
3. stick with our current introspection query, and do some ugly hack to get the column value. At the moment I'm thinking we can convert the row to JSON and then extract the relevant field from that; e.g.

```sql
select coalesce((row_to_json(pg_attribute) ->> 'attidentity'), '') as identity from pg_attribute limit 1;
```

3 is currently my preferred option.

Would you be interested in implementing this functionality?

**Comment by benjie** [comment]
I can't quite decide if this is a bug or not. Since it's on an input type, removing the not-null at a later date would not be a breaking change. So I'm going to stick with 'enhancement'

**Comment by wenzowski** [comment]
For the `GENERATED BY DEFAULT` case, I completely agree. It's the `GENERATED ALWAYS` that I was thinking makes it a bug instead of an enhancement–having the field available on the input type breaks expectations since any input value is discarded. I would expect anyone working with identity to understand that; someone working with the pure graphql schema...not so much.

Appreciate your detailed responses. I've got a few issues open that I'm happy to take ownership of. Might be a few weeks before I can get the PRs in though. Is there a branch tracking v4 somewhere or is the alpha package directly tracking #506 right now?

**Comment by benjie** [comment]
Yeah; #506 is the alpha. This issue is with `graphile-build`; once a new graphile-build release goes out I add it to #506 and publish the new postgraphql@next/postgraphile

**Comment by benjie** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [PG10 support (e.g. "generated always as identity")](https://trello.com/c/GBZ56aJ3/10-pg10-support-eg-generated-always-as-identity)

**Comment by benjie** [comment]
Support for this was released in v4.0.0-rc.5
